src/common/src/cast/mod.rs

```diff
@@ -163,70 +163,98 @@ pub fn i64_to_timestamp(t: i64) -> Result<Timestamp> {
+/// Refer to PostgreSQL's implementation <https://github.com/postgres/postgres/blob/5cb54fc310fb84287cbdc74533f3420490a2f63a/src/backend/utils/adt/varlena.c#L276-L288>
 pub fn str_to_bytea(elem: &str) -> Result<Box<[u8]>> {
-    // Padded with whitespace str is not allowed.
-    if elem.starts_with(' ') && elem.trim().starts_with("\\x") {
-        Err(PARSE_ERROR_STR_TO_BYTEA.to_string())
-    } else if let Some(remainder) = elem.strip_prefix(r"\x") {
+    if let Some(remainder) = elem.strip_prefix(r"\x") {
         Ok(parse_bytes_hex(remainder)?.into())
     } else {
         Ok(parse_bytes_traditional(elem)?.into())
     }
 }
 
-// Refer to Materialize: https://github.com/MaterializeInc/materialize/blob/1766ab3978bc90abf75eb9b1fbadfcc95eca1993/src/repr/src/strconv.rs#L623
-pub fn parse_bytes_hex(s: &str) -> Result<Vec<u8>> {
-    // Can't use `hex::decode` here, as it doesn't tolerate whitespace
-    // between encoded bytes.
+/// Ref: <https://docs.rs/hex/0.4.3/src/hex/lib.rs.html#175-185>
+fn get_hex(c: u8) -> Result<u8> {
+    match c {
+        b'A'..=b'F' => Ok(c - b'A' + 10),
+        b'a'..=b'f' => Ok(c - b'a' + 10),
+        b'0'..=b'9' => Ok(c - b'0'),
+        _ => Err(format!("invalid hexadecimal digit: \"{}\"", c as char)),
+    }
+}
 
-    let decode_nibble = |b| match b {
-        b'a'..=b'f' => Ok::<u8, String>(b - b'a' + 10),
-        b'A'..=b'F' => Ok(b - b'A' + 10),
-        b'0'..=b'9' => Ok(b - b'0'),
-        _ => Err(PARSE_ERROR_STR_TO_BYTEA.to_string()),
-    };
+/// Refer to <https://www.postgresql.org/docs/current/datatype-binary.html#id-1.5.7.12.10> for specification.
+pub fn parse_bytes_hex(s: &str) -> Result<Vec<u8>> {
+    let mut res = Vec::with_capacity(s.len() / 2);
 
-    let mut buf = vec![];
-    let mut nibbles = s.as_bytes().iter().copied();
-    while let Some(n) = nibbles.next() {
-        if let b' ' | b'\n' | b'\t' | b'\r' = n {
+    let mut bytes = s.bytes();
+    while let Some(c) = bytes.next() {
+        // white spaces are tolerated
+        if c == b' ' || c == b'\n' || c == b'\t' || c == b'\r' {
             continue;
         }
-        let n = decode_nibble(n)?;
-        let n2 = match nibbles.next() {
-            None => return Err(PARSE_ERROR_STR_TO_BYTEA.to_string()),
-            Some(n2) => decode_nibble(n2)?,
-        };
-        buf.push((n << 4) | n2);
+        let v1 = get_hex(c)?;
+
+        match bytes.next() {
+            Some(c) => {
+                let v2 = get_hex(c)?;
+                res.push((v1 << 4) | v2);
+            }
+            None => return Err("invalid hexadecimal data: odd number of digits".to_string()),
+        }
     }
-    Ok(buf)
+
+    Ok(res)
 }
 
-// Refer to https://github.com/MaterializeInc/materialize/blob/1766ab3978bc90abf75eb9b1fbadfcc95eca1993/src/repr/src/strconv.rs#L650
+/// Refer to <https://www.postgresql.org/docs/current/datatype-binary.html#id-1.5.7.12.10> for specification.
 pub fn parse_bytes_traditional(s: &str) -> Result<Vec<u8>> {
```

Comment:
I didn't have enough patience to rewrite this one to be Rusty, so it's just translated from C 🤡